proof/refine/ARM/Tcb_R.thy

```diff
@@ -1509,741 +1783,963 @@ lemma tc_corres_caps:
+                            simp: objBits_def objBitsKO_def valid_obj_size'_def
+                                  valid_sched_context_size'_def)
+          apply (fastforce elim: ex_cap_to'_after_update simp: ko_wp_at'_def tcb_cte_cases_def)
+         apply (clarsimp simp: idle_tcb'_2_def)
+        apply (clarsimp simp: valid_release_queue'_def obj_at'_def projectKO_eq projectKO_tcb)
+       apply (clarsimp simp: valid_release_queue'_def obj_at'_def projectKO_eq projectKO_tcb)
+      apply (clarsimp simp: untyped_ranges_zero_inv_def cteCaps_of_def comp_def)
+     apply simp
+    apply (fastforce simp: invs'_def valid_state'_def dest: global'_sc_no_ex_cap)
+   apply (clarsimp simp: state_relation_def invs_def valid_state_def valid_pspace_def)
+   apply (subgoal_tac "tcb_at' t s'")
+    apply (clarsimp simp: pspace_relation_def pred_tcb_at_def pred_tcb_at'_def obj_at_def obj_at'_def)
+    apply (drule_tac x=t in bspec; clarsimp simp: other_obj_relation_def tcb_relation_def projectKOs)
+   apply (fastforce elim: tcb_at_cross)
+  apply (clarsimp simp: sc_relation_def state_relation_def invs_def valid_state_def valid_pspace_def)
+  apply (subgoal_tac "sc_at' ptr s'")
+   apply (clarsimp simp: obj_at_def sc_at_pred_n_def obj_at'_def projectKOs pspace_relation_def)
+   apply (drule_tac x=ptr in bspec; clarsimp simp: sc_relation_def split: if_splits)
+  apply (fastforce simp: sc_at_pred_n_def obj_at_def is_sc_obj_def valid_obj_def elim!: sc_at_cross)
+  done
+
+schematic_goal finaliseSlot'_def:
+  "finaliseSlot' a b = ?X"
+  by (rule ext) simp
+
+lemma cteDelete_fh_lift:
+  assumes A: "\<And>x. \<lbrace>Q\<rbrace> emptySlot target x \<lbrace>\<lambda>_. P\<rbrace>"
+  and     B: "\<And>x. \<lbrace>R\<rbrace> cancelAllIPC x \<lbrace>\<lambda>_. Q\<rbrace>"
+  and     C: "\<And>s. (P and invs'  and L) s \<Longrightarrow> Q s \<and> R s"
+  shows "\<lbrace>P and invs' and cte_wp_at' (isValidFaultHandler \<circ> cteCap) target and L\<rbrace>
+         cteDelete target True
+         \<lbrace>\<lambda>_. P\<rbrace>"
+  apply (wpsimp wp: A simp: cteDelete_def)
+   prefer 2
+   apply assumption
+  apply (subst finaliseSlot_def)
+  apply (subst finaliseSlot'_def)
+  apply (rule hoare_vcg_seqE[rotated])
+   apply (subst liftE_validE)
+   apply (rule getCTE_sp)
+  apply (clarsimp split del: if_split)
+  apply (rule_tac Q="P and invs' and L and cte_wp_at' (\<lambda>c. c = cte) target
+                       and K (isValidFaultHandler (cteCap cte))" in hoare_pre(2))
+   apply (case_tac "cteCap cte"; clarsimp simp: isValidFaultHandler_def split: bool.splits)
+    apply (wpsimp simp: C)+
+         apply (rule hoare_FalseE)
+        apply (rule hoare_FalseE)
+       apply (rule hoare_FalseE)
+      apply (wpsimp wp: B isFinal simp: capRemovable_def finaliseCap_def isEndpointCap_def)+
+   apply (fastforce simp: C cte_wp_at'_def final_matters'_def
+                          capRemovable_def invs'_def valid_state'_def)+
+  done
+
+lemma hoare_vcg_common_conj_liftE:
```

Comment:
FIXME: move

Reply:
Can you please move this rather than labeling it?